### ISSUE TITLE
Algorithm yet to be implemented

### ISSUE THREAD

**Issue by Rajiv2605** [mlpack/mlpack#1112]
Hi there,
I am interested in implementing an algorithm or a feature in mlpack which hasn't been implemented yet. It would be great if you could suggest any :smile:

**Comment by zoq** [comment]
Hallo, maybe you can propose something that you like to see in mlpack. Another idea is to take a look at the GSoC ideas list: https://github.com/mlpack/mlpack/wiki/SummerOfCodeIdeas maybe you can find something interesting.

**Comment by Rajiv2605** [comment]
@zoq I'll look into it 😄

**Comment by Rajiv2605** [comment]
@zoq I found this to be interesting: https://github.com/mlpack/mlpack/wiki/SummerOfCodeIdeas#fast-k-centers-algorithm--implementation

**Comment by rcurtin** [comment]
I would strongly suggest finding a different project to pursue; that one is very difficult.  It would probably be best to select some machine learning algorithm that you do not see in mlpack that you think would be useful.

**Comment by Rajiv2605** [comment]
@rcurtin I went through the mplack repository and I don't think Convolutional Neural Network has been implemented yet. Can I work on it?

**Comment by zoq** [comment]
The Conv nets are implemented here: https://github.com/mlpack/mlpack/tree/master/src/mlpack/methods/ann; but if you interested in that direction, there is room for additions and improvements e.g. by implementing a faster conv layer. Let me know what you think.

**Comment by Rajiv2605** [comment]
@zoq Yes, I am interested! Could you tell me more about it?

**Comment by zoq** [comment]
One thing we could do is to remove unnecessary (temporary) copies inside the Forward and Backward function (https://github.com/mlpack/mlpack/blob/master/src/mlpack/methods/ann/layer/convolution_impl.hpp), like ``rotatedFilter`` in the ``Backward`` pass. Another idea is to implement a more efficient convolution operator by exploiting multicore CPU's e.g. http://proceedings.mlr.press/v70/budden17a.html could be interesting.

**Comment by Rajiv2605** [comment]
:+1: I'll look into it.

**Comment by zoq** [comment]
I saw your comment on the BigBatch SGD optimizer PR, if you are interested in implementing another optimizer, let me know there are a lot of interesting papers I really like to see implemented in the future like: YellowFin: https://arxiv.org/abs/1706.03471 or Natasha 2: https://arxiv.org/abs/1708.08694, etc.

**Comment by Rajiv2605** [comment]
@zoq I would love to implement an optimizer! YellowFin looks interesting. Could you suggest a few more?

**Comment by rcurtin** [comment]
Here's another one, CompAdaGrad: https://arxiv.org/pdf/1609.03319.pdf

However I won't lie, I have a secret motivation for that one---I know Nishant well (he actually wrote the LARS support and some other methods for mlpack), and I'd love to be able to tell him we implemented and tested his optimizer. :)  Maybe you are interested more by a different optimizer, so don't feel obligated to pick that one of course.

In any case, take a look at the existing optimizers that are there, and you should be able to implement whichever new technique you choose, then write some tests for it.  Once the code compiles and the tests pass, we can review and merge it.

**Comment by zoq** [comment]
I didn't know about CompAdaGrad, just skimmed over the paper, would be great to have an implementation.

**Comment by Rajiv2605** [comment]
@rcurtin CompAdaGrad looks interesting! I will work on it!

**Comment by Rajiv2605** [comment]
@rcurtin @zoq Hey, I will work on CompAdaGrad! This is the first time I am actually implementing a research paper code. So, it would be great if you could give many suggestions. How do I go about it? I am really excited about this and I believe that it would be a great learning experience.

**Comment by zoq** [comment]
Sorry for the slow response, what you could do is to go over an existing implementation/paper to see if you understand the overall idea; you could start with Adam or AdaGrad. If you have any questions, please feel free to join us on IRC #mlpack.

**Comment by Rajiv2605** [comment]
@zoq Thanks man! I'll do that :smile:

**Comment by sourabhvarshney111** [comment]
I would also like to implement any algorithm. Can I?

**Comment by zoq** [comment]
Of course, do you have something in mind? I think implementing an optimizer is a good start but we are always open for anything else.

**Comment by sourabhvarshney111** [comment]
@zoq @rcurtin Optimizing which algorithm?

**Comment by zoq** [comment]
That's up to you, but if you like you could look into: Nadam (Nesterov-accelerated Adaptive Moment Estimation), the optimizer combines Adam and NAG. Also, for examples you could take a look at: https://github.com/mlpack/mlpack/tree/master/src/mlpack/core/optimizers

**Comment by sourabhvarshney111** [comment]
Sorry for late response. But I would do that.

**Comment by haritha1313** [comment]
@zoq I have gone through many of the implemented optimizers and I think I will be able to implement nadam. I can't find any existing PR for the same. Can I move forward or is it already being worked upon?

**Comment by sourabhvarshney111** [comment]
@haritha1313 I am working upon it. I would complete it in a day or two.

**Comment by zoq** [comment]
@haritha1313 if you like you can take a look at: AMSGrad https://openreview.net/forum?id=ryQu7f-RZ and https://openreview.net/pdf?id=ryQu7f-RZ.

**Comment by haritha1313** [comment]
@zoq  ill definitely look into it :+1:  . Thanks for the suggestion.

**Comment by haritha1313** [comment]
@zoq I have implemented amsgrad and it is building successfully (optimizer files in mlpack/core/optimizers/ and the test file in mlpack/tests/). Is there any other standard procedure for testing implemented optimizers?

**Comment by zoq** [comment]
@haritha1313 That's it, we can solve and discuss any issues within the PR.

**Comment by gogameboy1** [comment]
@zoq Hi! I also want to implement one optimizer, do you have any suggestion? Thank you!

**Comment by zoq** [comment]
@gogameboy1 If you like you can take a look at: FTML: Follow the Moving Leader (http://www.cse.ust.hk/~szhengac/papers/icml17.pdf). Let me know if you think that might be something you like to work on.

**Comment by gogameboy1** [comment]
@zoq  Thank you! I will take a look on it.

**Comment by gogameboy1** [comment]
@zoq I would like to work on it, but I need some time to fully understand this paper. 
Thanks for this suggestion !

**Comment by sourabhvarshney111** [comment]
@zoq I want to work on new algorithm. Can you suggest any?

**Comment by zoq** [comment]
@sourabhvarshney111 there are a bunch of methods I like to see, maybe you are interested in implementing nesterov momentum? Don't feel obligated, just let me know.

**Comment by sourabhvarshney111** [comment]
@zoq Would look forward to find more on it.

**Comment by zoq** [comment]
@sourabhvarshney111 Okay great, let me know if you need any help.

**Comment by Awcrr** [comment]
@zoq hi Marcus, could you please suggest any other ML algorithm to implement? I'm free recently and  willing to contribute : )

**Comment by sourabhvarshney111** [comment]
@zoq Should I use this paper: https://arxiv.org/pdf/1607.01981.pdf for citing?

**Comment by zoq** [comment]
@sourabhvarshney111 I would go with:

```
@article{Nesterov1983,
    author  = {Nesterov, Yurii},
    journal = {Soviet Mathematics Doklady},
    pages   = {372--376},
    title   = {A method of solving a convex programming problem with
               convergence rate O(1/sqr(k))},
    volume  = {27},
    year    = {1983}
}
```

**Comment by zoq** [comment]
@Awcrr Happy to recommend something, do you like to implement another optimizer or something in another area? P.S.: nice picture.

**Comment by Awcrr** [comment]
@zoq optimizer is cool! Any recommendation?

**Comment by zoq** [comment]
@Awcrr If you like you can take a look at: FISTA (with backtracking line search): Paper: "A Fast Iterative Shrinkage-Thresholding Algorithm for Linear Inverse Problems" by A. Beck and M. Teboulle. Let us know if this is something you find interesting.

**Comment by sourabhvarshney111** [comment]
@zoq Thanks for the help. I would go through it.

**Comment by Awcrr** [comment]
@zoq Thanks very much for the reference. Just took a look at this paper. It's great. I think I will read it more carefully and have a try.

**Comment by nvinayvarma189** [comment]
Hello @zoq @rcurtin ,
I've been trying to make my first contribution for a long time and I've been following your conversations. Can you help me with implementing a ML algorithm or an optimiser. I will try my best to cope up with you people

**Comment by zoq** [comment]
Hello @nvinayvarma189 if you like you could look into the RPROP method as proposed by M. Riedmiller et al. in "Rprop - A Fast Adaptive Learning Algorithm". Another interesting idea might be: Projected gradient descent for non-negative matrix factorization: C.-J. Lin. "Projected gradient methods for non-negative matrix factorization". Let us know what you think.

**Comment by nvinayvarma189** [comment]
Thank you @zoq , both of them look good to me . can you tell me more on how to present my algorithm?

**Comment by rcurtin** [comment]
@nvinayvarma189: take a look at some of the other PRs that have been opened for optimizers (like #1168 for instance, and there are others too) and this should help give you an idea of what to do.

**Comment by nvinayvarma189** [comment]
Thanks @rcurtin . I would definitely look into and will try to open a PR asap.

**Comment by SwapnilDreams100** [comment]
Hi, is any variant of GAN or LSTM to be implemented? Would love to work on it, since I have python experience with them but c++ implementation would be refreshing! Otherwise any optimiser looks good too.

**Comment by zoq** [comment]
Hello @SwapnilDreams100, there is already an LSTM implementation (https://github.com/mlpack/mlpack/blob/master/src/mlpack/methods/ann/layer/lstm.hpp), and there is an open PR for standard GAN (https://github.com/mlpack/mlpack/pull/1066), which needs some work. If you like you can take a look at the PR and work on that. Another option is to take a look at regularizers, there is a thread on the mailing list with additional information: http://knife.lugatgt.org/pipermail/mlpack/2018-January/003447.html

**Comment by RohanRajadhyax** [comment]
@zoq @rcurtin I am interested in implementing any of the deep learning algorithms or a feature in mlpack.
Any suggestion for your end would be helpful.

**Comment by zoq** [comment]
Hello @RohanRajadhyax, one feature that is currently missing is ``arma::cube`` support for the ``Binarize`` method (https://github.com/mlpack/mlpack/blob/master/src/mlpack/core/data/binarize.hpp). Currently, it just works for ``arma::mat`` so what we currently do is to go through the object slice by slice; here is an example: https://github.com/mlpack/mlpack/blob/master/src/mlpack/tests/recurrent_network_test.cpp#L706-L711

So in the best case, we could templatize the existing implementation so that it works for ``arma::mat`` as well as for ``arma::cube``. Let us know if that is something you like to work on.

**Comment by deepakks1995** [comment]
Hi @zoq , I would like to know how you are planning to templatize the `arma::code` and `arma::mat` in the Binarize method. One solution which I think is that  we take the input of binarize function as a `arma::cube` and the process the input slice by slice in the Binarize function and then return the output as a cube. Correct me if I am wrong

**Comment by zoq** [comment]
```c++
template<typename T>
void Binarize(const arma::Mat<T>& input,
              arma::Mat<T>& output,
              const double threshold)
{
  output.copy_size(input);

  const T *inPtr = input.memptr();
  T *outPtr = output.memptr();

  #pragma omp parallel for
  for (omp_size_t i = 0; i < (omp_size_t) input.n_elem; ++i)
    outPtr[i] = inPtr[i] > threshold;
}
```

This is the first method of the two, which should work just fine if we use ``MatType`` instead of ``arma::Mat<T>``.


The second function isn't that simple, in this case, I think we could provide a specialized implementation for ``arma::Cube<T>`` and proceed as you suggested.

**Comment by PranayMukherjee** [comment]
I would really love to add something new to mlpack. Could you guys suggest some possible algos. which havent been implemented yet?

**Comment by Prabhat-IIT** [comment]
Hi @zoq ....(dungeon_master here 😅) can I work on arma::cube feature or any other optimiser or any new deep learning algorithm?

**Comment by zoq** [comment]
@PranayMukherjee @Prabhat-IIT one idea would be to implement the Glorot initialization method (http://proceedings.mlr.press/v9/glorot10a/glorot10a.pdf), here are the already implemented methods that could be helpful: https://github.com/mlpack/mlpack/tree/master/src/mlpack/methods/ann/init_rules

Another idea is to implement an optimization method e.g. SAGA (https://papers.nips.cc/paper/5258-saga-a-fast-incremental-gradient-method-with-support-for-non-strongly-convex-composite-objectives.pdf), in this case, someone could take a look at the existing optimization algorithms:
 https://github.com/mlpack/mlpack/tree/master/src/mlpack/core/optimizers

**Comment by Prabhat-IIT** [comment]
@zoq  So we have to just implement the 'Normalized Initiliazation' part of the Glorot Initialization. A method giving initial weights according to the proposed soln in the paper would suffice I think?

**Comment by zoq** [comment]
@Prabhat-IIT Correct.

**Comment by zoq** [comment]
I'm closing this issue, ideally, anyone who likes to contribute a new method should identify something that mlpack is currently missing and start a discussion e.g. here on Github, IRC or via the mailing list (see http://www.mlpack.org/help.html for more information). We are more than happy to answer any questions and help with any upcoming issues.
